packages/openapi3/src/xml-module.ts

```diff
@@ -0,0 +1,231 @@
+  resolveEncodedName,
+} from "@typespec/compiler";
+import { ArrayBuilder, ObjectBuilder } from "@typespec/compiler/emitter-framework";
+import { reportDiagnostic } from "./lib.js";
+import { ResolvedOpenAPI3EmitterOptions } from "./openapi.js";
+import { getSchemaForStdScalars } from "./std-scalar-schemas.js";
+import { OpenAPI3Schema, OpenAPI3XmlSchema } from "./types.js";
+
+export interface XmlModule {
+  attachXmlObjectForScalarOrModel(
+    program: Program,
+    type: Scalar | Model,
+    emitObject: OpenAPI3Schema,
+  ): void;
+
+  attachXmlObjectForModelProperty(
+    program: Program,
+    options: ResolvedOpenAPI3EmitterOptions,
+    prop: ModelProperty,
+    emitObject: OpenAPI3Schema,
+    ref?: Record<string, any>,
+  ): void;
+}
+
+export async function resolveXmlModule(): Promise<XmlModule | undefined> {
+  const xml = await tryImportXml();
+  if (xml === undefined) return undefined;
+
+  return {
+    attachXmlObjectForScalarOrModel: (
+      program: Program,
+      type: Scalar | Model,
+      emitObject: OpenAPI3Schema,
+    ) => {
+      const isXmlModel = isXmlModelChecker(program, type, []);
+      if (!isXmlModel) {
+        return;
+      }
+
+      const xmlObject: OpenAPI3XmlSchema = {};
+
+      // Resolve XML name
+      const xmlName = resolveEncodedName(program, type, "application/xml");
+      if (xmlName !== type.name) {
+        xmlObject.name = xmlName;
+      }
+
+      // Get and set XML namespace if present
+      const currNs = xml.getNs(program, type);
+      if (currNs) {
+        xmlObject.prefix = currNs.prefix;
+        xmlObject.namespace = currNs.namespace;
+      }
+
+      // Attach xml schema to emitObject if not empty
+      if (Object.keys(xmlObject).length !== 0) {
+        emitObject.xml = xmlObject;
+      }
+    },
+    attachXmlObjectForModelProperty: (
+      program: Program,
+      options: ResolvedOpenAPI3EmitterOptions,
+      prop: ModelProperty,
+      emitObject: OpenAPI3Schema,
+      refSchema: OpenAPI3Schema,
+    ) => {
+      const xmlObject: OpenAPI3XmlSchema = {};
+
+      const isXmlModel = isXmlModelChecker(program, prop.model!, []);
+      if (!isXmlModel) {
+        return;
+      }
+
+      // Resolve XML name
+      const xmlName = resolveEncodedName(program, prop, "application/xml");
+      const jsonName = resolveEncodedName(program, prop, "application/json");
+      if (xmlName !== prop.name && xmlName !== jsonName) {
+        xmlObject.name = xmlName;
+      }
+
+      // Get and set XML namespace if present
+      const currNs = xml.getNs(program, prop);
+      if (currNs) {
+        xmlObject.prefix = currNs.prefix;
+        xmlObject.namespace = currNs.namespace;
+      }
+
+      // Set XML attribute if present
+      const isArrayProperty = prop.type?.kind === "Model" && isArrayModelType(program, prop.type);
+      const isAttribute = xml.isAttribute(program, prop);
+      if (isAttribute) {
+        if (prop.type?.kind === "Model" && isArrayProperty) {
```

Comment:
why it is a special case to have that for arrays vs other structure? I think just make it invalid warning and default to a string in all those cases.

Reply:
you haven't responded to this comment...